### PR TITLE
TINY-8351: RTC doesn't fire all relevant get/set content events

### PR DESCRIPTION
**Pull request by vpyshnenko**
Related Ticket: TINY-8351

Description of Changes:
* The main goal of this PR is to unify the behavior about sending events on `getContent`, `setContent`, and `insertContent` calls by wrapping these methods with `preProcess(Get/Set)Content` and `postProcess(Get/Set)Content` functions.
* Having this logic in the one place should cover internal `tinymce` methods implementation as well as correspondent RTC's one.
* Move out any fire pre/post events logic from RTC to the `tinymce` core.

Pre-checks:
* [x] ~Changelog entry added~
* [x] ~Tests have been added (if applicable)~
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Review by TheSpyder — APPROVED** [review]

**Review by lnewson — CHANGES_REQUESTED** [review]
This isn't what was mentioned in the Jira and now adds more duplication which isn't what we want. We need to change where the RTC integration hooks in, not duplicate the logic in here.

So to be clear, the `SetContent.ts` file should contain the event processing logic and then it should do the RTC adapter delegation from there.

**Review by TheSpyder — CHANGES_REQUESTED** [review]
I was wondering about that, but I assumed it been discussed while I was away.

**Comment by lnewson** [comment]
> I was wondering about that, but I assumed it been discussed while I was away.

As far as I'm aware nothing extra was discussed... I suspect the `insertContent` one could be difficult, in which case this is probably going to be okay if we run out of time. Ideally I'd like to see us avoid that for everything though and only do the current approach should we be short on time.

File: modules/tinymce/src/core/main/ts/Rtc.ts
Comment (TheSpyder) [inline modules/tinymce/src/core/main/ts/Rtc.ts:232]: I like the changes you've done with set and get content, please do the same with insert content

- [ ] move insert content processing out of `Rtc.ts`

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/content/ContentTypes.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/ContentTypes.ts:32]:
```suggestion
  readonly content: Content;
  readonly html: string;
```

As this appears to be an internal only type we should make sure the properties are readonly as well.

File: modules/tinymce/src/core/main/ts/content/SetContent.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/SetContent.ts:34]: Why is this called `_content`? That naming scheme is only meant to be used for unused attributes, but it's definitely used here. Maybe instead of destructuring above, do something like this?
```
    const result = Rtc.setContent(editor, updatedContent, updatedArgs);
    postProcessSetContent(editor, result.html, updatedArgs);
    return result.content;
```

File: modules/tinymce/src/core/main/ts/content/GetContentImpl.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/GetContentImpl.ts:43]:
```suggestion
  return shouldTrim && Type.isString(content) ? Tools.trim(content) : content;
```

File: modules/tinymce/src/core/main/ts/content/SetContent.ts
Comment (vpyshnenko) [reply modules/tinymce/src/core/main/ts/content/SetContent.ts:34]: Initially, I wanted to avoid using the additional variable name and confusing associations with the` Result` type. But destruction to {content, html} ends up to strange error for the previous line in code :
`error TS2448: Block-scoped variable 'content' used before its declaration.`
 even though the `content` argument looks visible in scope.

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/content/SetContent.ts:34]: It's possibly related to the fact that `content` is already in scope. Either way, destructuring isn't really getting any benefit here and I personally don't think it's confusing with the `Result` type. However, if you're worried about that then we should just come up with another name instead... maybe `contentResult` or `setContentResult` to match the `SetContentResult` type?

Comment (vpyshnenko) [reply modules/tinymce/src/core/main/ts/content/SetContent.ts]: I've reverted to the initially suggested variant without destructuring

File: modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts:225]: There's a few cleanup steps that are run before here, should they also be included in `InsertContent.ts`? (mainly just looking at the whitespace cleanup)

File: modules/tinymce/src/core/main/ts/content/InsertContent.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/InsertContent.ts:49]: This is incorrect... the `Rtc.insertContent` function needs to return the processed HTML and pass that to `postProcessSetContent` as that's what was passed previously here, not the input HTML.

**Review by lnewson — CHANGES_REQUESTED** [review]
Looking good and almost there, but the content being fired in the `SetContent` event when inserting HTML is now incorrect/different. So we need that fixed before we can merge this sorry.

File: modules/tinymce/src/core/main/ts/content/InsertContent.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/InsertContent.ts:48]: This line is also incorrect, as we should be passing the pre-processed HTML, not the input HTML.
```suggestion
    Rtc.insertContent(editor, args.content, details);
```

File: modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts
Comment (TheSpyder) [inline modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts:264]: Was this here incorrectly? Or should it be replaced with firing an event somewhere else?

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts:264]: It was moved to the `InsertContent.ts` file from what I saw, as it's handled by the `postProcessSetContent` call now I believe?

File: modules/tinymce/src/core/main/ts/content/InsertContent.ts
Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/content/InsertContent.ts:49]: Ah oops. The plugin has never done that.

For set content RTC just re-serializes the HTML model returned by the TinyMCE parser, but that's still worth doing the same here rather than just using the input string. The parser will address things like security issues.

File: modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts
Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts:264]: Ah I didn't realise it was fired elsewhere already. It just seemed weird that the `fire` was removed without another added elsewhere in the PR.

File: modules/tinymce/src/core/main/ts/content/InsertContent.ts
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/content/InsertContent.ts:61]:
```suggestion
    const insertedContent = Rtc.insertContent(editor, args.content, details);
    postProcessSetContent(editor, insertedContent, args);
```
Maybe? It just looked a little overloaded 🤷

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/content/InsertContent.ts
Comment (spocke) [reply modules/tinymce/src/core/main/ts/content/InsertContent.ts]: Yeah, I just passed out the serialized input ast to get a value to send to the events setContent event I guess you should serialize the alt model fragment using the alt model serializer to get the lists and elements flattning like the actual inserted value. However the whole purpose from what I can see of the setContent event is to do things to the dom when contents gets inserted.

Not sure what the real need would be exactly to get the html being inserted after it has been inserted you can't really alter it at that point that is what the `beforeSetContent` event is for but all this was added before there where proper node filters etc. The events here makes less sense in the alt model. That is why I was not firing all of them.

If you look at every instance we use the `setContent` event it's not even caring about the event data it just does dom mutation or dom reading after some insert is made. So maybe we should deprecate that `content` property from the `setContent` event to me it serves little purpose. Also the serialization process is currently a bit expensive in the alt model since we unflatten marks, unflatten lists etc.

**Review by spocke — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/content/InsertContent.ts
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/content/InsertContent.ts]: No objections to deprecating it here. For RTC you could likely just return an empty string in that case.

Comment (spocke) [reply modules/tinymce/src/core/main/ts/content/InsertContent.ts]: I added a ticket for that #TINY-8457 since those changes would hold up this one.

**Review by spocke — APPROVED** [review]

**Pull request merged**
